### PR TITLE
Add auth token for Python template

### LINKED ISSUE #1173: [Python] create azd credential authn handler

### PR DESCRIPTION
**Pull request by hemarina**
Fix #1173

### CODE REVIEW

File: templates/todo/projects/python-mongo-swa-func/.repo/bicep/repo.yaml [hunk 1]
```diff
@@ -145,6 +145,10 @@ repo:
     - from: ../../../../common/tests
       to: ./tests
 
+    # Auth Python common
+    - from: ../../../../common/auth/py
+      to: ./src/api/src
```
Comment (vhvb1989) [inline templates/todo/projects/python-mongo-swa-func/.repo/bicep/repo.yaml:150]:
```suggestion
      to: ./src/api/todo
```

File: templates/todo/common/auth/py/AzureDeveloperCliCredential.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+def parse_token(output):
+    try:
+        token = json.loads(output)
+        dt = datetime.strptime(token["expiresOn"], "%Y-%m-%dT%H:%M:%SZ")
+        if hasattr(dt, "timestamp"):
+            # Python >= 3.3
+            expires_on = dt.timestamp()
+        else:
+            # taken from Python 3.5's datetime.timestamp()
+            expires_on = time.mktime((dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second, -1, -1, -1))
```
Comment (vhvb1989) [inline templates/todo/common/auth/py/AzureDeveloperCliCredential.py:60]:
```suggestion
        expires_on = dt.timestamp()
```

Azd has a dependency on Python version 3.7:

image

So, we don't need this condition to support python <= 3.3

File: templates/todo/common/auth/py/AzureDeveloperCliCredential.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        working_directory = get_safe_working_dir()
+
+        kwargs = {
+            "stderr": subprocess.PIPE,
+            "cwd": working_directory,
+            "universal_newlines": True,
+            "env": dict(os.environ, NO_COLOR="true"),
+        }
+        if platform.python_version() >= "3.3":
+            kwargs["timeout"] = 10
```
Comment (vhvb1989) [inline templates/todo/common/auth/py/AzureDeveloperCliCredential.py:97]: Same, we don't need to check for py 3.3 support here.

```suggestion
        kwargs = {
            "stderr": subprocess.PIPE,
            "cwd": working_directory,
            "universal_newlines": True,
            "env": dict(os.environ, NO_COLOR="true"),
            "timeout": 10,
        }
```

**Review by vhvb1989 — APPROVED** [review]
A few suggestions but LGTM.

Consider using `src/api/todo` folder instead of adding the `src/api/src` folder just with one file.
It looks like all the python code is inside the `todo` folder only.

File: templates/todo/common/auth/py/azureDeveloperCliCredential.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            message = "Failed to invoke Azure Developer CLI"
+        raise ClientAuthenticationError(message=message)
+    except OSError as ex:
+        # failed to execute 'cmd' or '/bin/sh'; CLI may or may not be installed
+        error = CredentialUnavailableError(message="Failed to execute '{}'".format(args[0]))
+        six.raise_from(error, ex)
+    except Exception as ex:  # pylint:disable=broad-except
+        # could be a timeout, for example
+        error = CredentialUnavailableError(message="Failed to invoke the Azure Developer CLI")
+        six.raise_from(error, ex)
```
Comment (vhvb1989) [inline templates/todo/common/auth/py/azureDeveloperCliCredential.py:114]:
```suggestion
        six.raise_from(error, ex)

```

Comment (vhvb1989) [reply templates/todo/common/auth/py/azureDeveloperCliCredential.py:114]: missing new line

**Review by wbreza — APPROVED** [review]

**Pull request merged**